Switch mainboard code and bootloader to run on new hardware

timburke opened the pull request:
The next-generation momo hardware uses a PIC24FJ64GA306 chip as its MPU.  This is decently compatible with the PIC24F16KA101 but has some differences.  The mainboard code and bootloader have been ported to the new hardware and this pull request tracks their status.

_It should not be merged yet_ but only after integrating the sensor_log and sleep code into branch.

This PR is to show where the code is.  Bootloading and mainboard code is now working on the new chip.

Major hw changes are:
- Flash is 64KB vs 16KB
- RAM is 8KB vs 1.5KB
- IO pin assignment to peripherals is done through "Peripheral Pin Select" so there are not fixed pins assigned to the SPI port for example.  This is not true of I2c which for some reason does still have fixed pins.
- There is no EEPROM
- There are many more IO pins

Software changes:
- bootloader updated to not use EEPROM since there is none
- SPI code is reclocked to run at full-speed since there was no reason for it to run at 1/512th speed as had been the case until now.
- ioport.h file has macros now for manipulating ports.
- Updated code to use slightly different SPI registers in new chip.
- Made pic24 library retargetable between this chip and the previous 24F16KA101.

timburke committed:
Update bootloader to work with new board.

- now waits for ALARM pin to be released before starting a reflash.
This fixes a bug with using Pickit3 to program the bootloader where
the bootloader would trigger immediately after being programmed
- updated SPI flash to full-speed.  There was no reason to be using
it at 1/512th speed.
- communication with main controller firmware is now done not with
EEPROM but a perisistent global variable.  The new chip has no EEPROM

timburke committed:
Update shared pic24 module for new PIC.

- Code paths that must be different on Pic24F and Pic24FJ are now
defined using the preprocessor for each chip.
- The ADC and SPI flash code has been updated.
- Upgraded SPI flash speed to full-speed. Things are *much* faster
now.  We had been using it at the (default) 1/512th speed until now.

timburke committed:
Update controller code for new board.

- RTCC lives forever now with the backup battery, so
don't reset it on a power-on reset
- reflash endpoint now uses shared global variable to communicate
with bootloader rather than flash.
- updated linker script with new memory map for new chip

timburke committed:
Add macros for easy IOPort manipulation

timburke committed:
Add support for targeting multiple chips with pic24 lib.

- rake buildFJ will target the PIC24FJ64GA306
- rake buildF will target he PIC24F16KA101

There is no default build since you need to pick which one you
want to target.

timburke committed:
Update FSU to use correctly targeted library

timburke committed:
Update mainboard to use correctly targeted libPic24

timburke committed:
Remove #error directive when target is not PIC24F16KA101

timburke committed:
Merge branch 'pr-controller-sleep' into pr-support-pic24FJ64GA306

Conflicts:
	momo_modules/mainboard/src/main.c
	momo_modules/mainboard/src/mib/controller_mib_feature.c
	momo_modules/shared/pic24/src/modules/memory.c

Merge in controller sleep functionality with new chip code.

timburke committed:
Remove remaining merge artifact

timburke committed:
Sleep by default

timburke commented:
Controller sleep code is now integrated and tested on the new chip.  Controller board sleep current draw: 13 uA.

timburke committed:
Add light sleep option to keep Flash memory hot.

Normally the flash memory is powerd down during sleep to conserve
power but when we are pushing firmware, the wakeup time to bring
the flash back up is ~10 ms per 20 byte push RPC since we have time
to sleep after each one.  This makes pushing slower than it needs
to be.

This commit adds a flag that tells the sleep_handler to not power
down the flash memory if we expect to wake up and use it very soon

amcgee commented:
13 uA, nice!!

I think we should either:
1) tag/branch the repo before integrating this, so we have a snapshot we can go back to that will compile and run on the older chips (of which there are a few in the wild) OR
2) make the build process (so mainboard_bootloader and mainboard) create executables for each chip (like we do with the 8bit guys)

timburke committed:
Update pic12 executive to not boot apps on ALARM

You can now force all of the mib modules to not start
running user code if ALARM is low on reset.

timburke committed:
Remove potential bug with analog pins.

You should always set a pin level using LATXY, not RXY.
Internally the cpu does a read-modify-write operation and
analog pins are read as 0, so if you touch RXY, you can
unintentionaly reset all the analog pins to drive 0s.  LATX
if the register holding the output values, RX is the register
holding the currently measured values.

timburke committed:
Fix bus in bus_reset code.

The bus reset line is high for cutting the bus off, not low.

timburke committed:
Add power on rpc call

timburke committed:
Merge branch 'dev' of github.com:WellDone/MoMo-Firmware into pr-support-pic24FJ64GA306

timburke committed:
Merge branch 'dev' into pr-support-pic24FJ64GA306

Conflicts:
	momo_modules/mainboard/src/mib/controller_mib_feature.h

Merge sensor api code from latest dev and fix merge errors.

timburke commented:
I have integrated the sensor api into this branch and tested momosensor on the new hardware.  It works well and it lovely!!  I created a tag pointing to the last commit on dev so that we can always go back to the old board code if we need.

The PIC library code is multi-targeted.  We could update the pic24 build system to use scons and be multitargeted, which is probably a good thing to do anyway to get off of ceedling, but i'm not sure it's worth the effort right now given other priorities.

timburke committed:
Update ADC clock to reduce noise

amcgee commented on momo_modules/mainboard/src/main.c:
```diff
@@ -8,59 +8,52 @@
+#pragma config POSCMD = NONE            // Primary Oscillator Select (Primary Oscillator Disabled)
+#pragma config BOREN1 = EN              // BOR Override bit (BOR Enabled [When BOREN=1])
+#pragma config IOL1WAY = OFF            // IOLOCK One-Way Set Enable bit (The IOLOCK bit can be set and cleared using the unlock sequence)
+#pragma config OSCIOFNC = ON            // OSCO Pin Configuration (OSCO/CLKO/RC15 functions as port I/O (RC15))
```
So apparently this doesn't work with xc16-gcc version 1.21.  I get the error:
`src/main.c:32:9: error: unknown configuration setting: 'OSCIOFNC'`

timburke committed:
Update main.c

Fix typo from mplabx in config name

timburke committed:
Update battery.c

Fix overflow since autosample_wait is only 5 bits.
